cpp/src/knn/knn_opg_common.cu

```diff
@@ -392,366 +273,413 @@ void perform_local_knn(int64_t *res_I, float *res_D,
   }
 }
 
-template <typename T>
-void opg_knn(raft::handle_t &handle, std::vector<Matrix::Data<T> *> *out,
-             std::vector<Matrix::Data<int64_t> *> *out_I,
-             std::vector<Matrix::floatData_t *> *out_D,
-             std::vector<Matrix::floatData_t *> &idx_data,
-             Matrix::PartDescriptor &idx_desc,
-             std::vector<Matrix::floatData_t *> &query_data,
-             Matrix::PartDescriptor &query_desc,
-             std::vector<std::vector<T *>> &y, bool rowMajorIndex,
-             bool rowMajorQuery, int k, int n_outputs, size_t batch_size,
-             bool verbose, std::vector<std::vector<float *>> *probas = nullptr,
-             std::vector<int *> *uniq_labels = nullptr,
-             std::vector<int> *n_unique = nullptr, bool probas_only = false) {
-  ASSERT(k <= 1024, "k must be <= 1024");
-  ASSERT(batch_size > 0, "max_batch_size must be > 0");
-  ASSERT(k < idx_desc.M, "k must be less than the total number of query rows");
-  for (Matrix::RankSizePair *rsp : idx_desc.partsToRanks) {
-    ASSERT(rsp->size >= k,
-           "k must be <= the number of rows in the smallest index partition.");
-  }
-
-  const raft::handle_t &h = handle;
-  const auto &comm = h.get_comms();
-  cudaStream_t stream = h.get_stream();
-
-  const auto allocator = h.get_device_allocator();
+void reduce(opg_knn_param &params, opg_knn_utils &utils, cuda_utils &cutils,
+            int part_idx, size_t processed_in_part, size_t batch_size) {
+  device_buffer<int64_t> trans(cutils.alloc, cutils.stream,
+                               utils.idxRanks.size());
+  CUDA_CHECK(cudaMemsetAsync(
+    trans.data(), 0, utils.idxRanks.size() * sizeof(int64_t), cutils.stream));
 
-  int my_rank = comm.get_rank();
+  size_t batch_offset = processed_in_part * params.k;
 
-  std::set<int> idxRanks = idx_desc.uniqueRanks();
-
-  std::vector<Matrix::RankSizePair *> local_idx_parts =
-    idx_desc.blocksOwnedBy(comm.get_rank());
-
-  int local_parts_completed = 0;
-
-  // Loop through query parts for all ranks
-  for (int i = 0; i < query_desc.totalBlocks(); i++) {
-    Matrix::RankSizePair *partition = query_desc.partsToRanks[i];
-    int part_rank = partition->rank;
-    size_t part_n_rows = partition->size;
-
-    size_t total_batches = raft::ceildiv(part_n_rows, batch_size);
-    size_t total_n_processed = 0;
-
-    // Loop through batches for each query part
-    for (int cur_batch = 0; cur_batch < total_batches; cur_batch++) {
-      size_t cur_batch_size = batch_size;
-
-      if (cur_batch == total_batches - 1)
-        cur_batch_size = part_n_rows - (cur_batch * batch_size);
-
-      if (my_rank == part_rank) CUML_LOG_DEBUG("Root Rank is %d", my_rank);
-
-      /**
-       * Root broadcasts batch to all other ranks
-       */
-      CUML_LOG_DEBUG("Rank %d: Performing Broadcast", my_rank);
-
-      int my_rank = comm.get_rank();
-      device_buffer<float> part_data(allocator, stream, 0);
+  char32_t *outputs = nullptr;
```

Comment:
Yes, I used `char32_t` for (ints/classification and floats/regression) as a way to avoid templates in order to speed up compilation and to get a lighter binary. This was made possible by the fact that no operation is made with these values apart from reads and writes. However, as you said, it sacrifices the ability to use other types. I switched to using templates.